File: lib/db/library_repository.dart
import 'package:sembast/sembast.dart';
import '../models/book.dart';
import '../models/bookmark.dart';
import 'sembast_db.dart';

class LibraryRepository {
  Future<void> addBook(Book book) async {
    final db = await SembastDb.instance();
    await SembastDb.booksStore.record(book.uuid).put(db, book.toMap());
  }

  Future<void> updateBook(Book book) async {
    final db = await SembastDb.instance();
    await SembastDb.booksStore.record(book.uuid).put(db, book.toMap());
  }

  Future<void> deleteBook(String uuid) async {
    final db = await SembastDb.instance();
    await SembastDb.booksStore.record(uuid).delete(db);
  }

  Future<Book?> getBook(String uuid) async {
    final db = await SembastDb.instance();
    final record = await SembastDb.booksStore.record(uuid).get(db);
    return record == null ? null : Book.fromMap(record);
  }

  Future<List<Book>> allBooks() async {
    final db = await SembastDb.instance();
    final records = await SembastDb.booksStore.find(
      db,
      finder: Finder(sortOrders: [SortOrder('dateAdded', false)]),
    );
    return records.map((r) => Book.fromMap(r.value)).toList();
  }

  Future<List<Book>> booksByStatus(ReadingStatus status) async {
    final db = await SembastDb.instance();
    final records = await SembastDb.booksStore.find(
      db,
      finder: Finder(
        filter: Filter.equals('status', status.index),
        sortOrders: [SortOrder('title')],
      ),
    );
    return records.map((r) => Book.fromMap(r.value)).toList();
  }

  /// Real-time instant search across title, author, and genre tags —
  /// runs in-memory over the (small, personal-library-sized) book set.
  Future<List<Book>> search(String queryText) async {
    final all = await allBooks();
    if (queryText.trim().isEmpty) return all;
    final q = queryText.toLowerCase();
    return all.where((b) {
      if (b.title.toLowerCase().contains(q)) return true;
      if (b.author.toLowerCase().contains(q)) return true;
      if (b.genreNames.any((g) => g.toLowerCase().contains(q))) return true;
      return false;
    }).toList();
  }

  Future<List<String>> allGenreNames() async {
    final all = await allBooks();
    final set = <String>{};
    for (final b in all) {
      set.addAll(b.genreNames);
    }
    return set.toList()..sort();
  }

  // ---- Bookmarks ----

  Future<void> addBookmark(Bookmark bookmark) async {
    final db = await SembastDb.instance();
    await SembastDb.bookmarksStore.record(bookmark.id).put(db, bookmark.toMap());
  }

  Future<List<Bookmark>> bookmarksForBook(String bookUuid) async {
    final db = await SembastDb.instance();
    final records = await SembastDb.bookmarksStore.find(
      db,
      finder: Finder(filter: Filter.equals('bookUuid', bookUuid)),
    );
    return records.map((r) => Bookmark.fromMap(r.value)).toList();
  }

  Future<void> deleteBookmark(String id) async {
    final db = await SembastDb.instance();
    await SembastDb.bookmarksStore.record(id).delete(db);
  }
}

File: src/db/libraryRepository.ts
import { db } from './database';
import type { Book, BookFile, Bookmark, ReadingStatus } from '../types';

/**
 * Single point of access for all local persistence. Every method
 * here reads/writes IndexedDB (via Dexie) only.
 */
export const libraryRepository = {
  async addBook(book: Book, file: BookFile): Promise<void> {
    await db.transaction('rw', db.books, db.bookFiles, async () => {
      await db.books.put(book);
      await db.bookFiles.put(file);
    });
  },

  async updateBook(book: Book): Promise<void> {
    await db.books.put(book);
  },

  async deleteBook(uuid: string): Promise<void> {
    await db.transaction('rw', db.books, db.bookFiles, db.bookmarks, async () => {
      await db.books.delete(uuid);
      await db.bookFiles.delete(uuid);
      await db.bookmarks.where('bookUuid').equals(uuid).delete();
    });
  },

  async getBook(uuid: string): Promise<Book | undefined> {
    return db.books.get(uuid);
  },

  async getBookFile(uuid: string): Promise<BookFile | undefined> {
    return db.bookFiles.get(uuid);
  },

  async allBooks(): Promise<Book[]> {
    return db.books.orderBy('dateAdded').reverse().toArray();
  },

  async booksByStatus(status: ReadingStatus): Promise<Book[]> {
    return db.books.where('status').equals(status).sortBy('title');
  },

  /** Real-time instant search across title, author, and genre tags. */
  async search(queryText: string): Promise<Book[]> {
    const all = await this.allBooks();
    const q = queryText.trim().toLowerCase();
    if (!q) return all;
    return all.filter(
      (b) =>
        b.title.toLowerCase().includes(q) ||
        b.author.toLowerCase().includes(q) ||
        b.genreNames.some((g) => g.toLowerCase().includes(q)),
    );
  },

  async allGenreNames(): Promise<string[]> {
    const all = await this.allBooks();
    const set = new Set<string>();
    for (const b of all) b.genreNames.forEach((g) => set.add(g));
    return [...set].sort();
  },

  // ---- Bookmarks ----

  async addBookmark(bookmark: Bookmark): Promise<void> {
    await db.bookmarks.put(bookmark);
  },

  async bookmarksForBook(bookUuid: string): Promise<Bookmark[]> {
    return db.bookmarks.where('bookUuid').equals(bookUuid).toArray();
  },

  async deleteBookmark(id: string): Promise<void> {
    await db.bookmarks.delete(id);
  },
};
